### PR TITLE
Fix Python 3.14 compatibility for typing.Union annotations

### LINKED ISSUE #2064: Python 3.14 compatibility
In Python 3.14, some implementation details of typing have been changed, for example: https://github.com/python/cpython/issues/105499 

This is causing failures of some of the tests:

```
=================================== FAILURES ===================================
________________ test_string_annotation[annotations8-result8-] _________________

annotations = {'return': 'typing.Union[str, int]'}, result = ['int', 'str']
code = ''

    @pytest.mark.parametrize(
        'annotations, result, code', [
            ({}, [], ''),
            (None, [], ''),
            ({'asdf': 'str'}, [], ''),
    
            ({'return': 'str'}, ['str'], ''),
            ({'return': 'None'}, ['NoneType'], ''),
            ({'return': 'str().upper'}, [], ''),
            ({'return': 'foo()'}, [], ''),
            ({'return': 'bar()'}, ['float'], ''),
    
            # typing is available via globals.
            ({'return': 'typing.Union[str, int]'}, ['int', 'str'], ''),
            ({'return': 'typing.Union["str", int]'},
             ['int', 'str'] if sys.version_info >= (3, 9) else ['int'], ''),
            ({'return': 'typing.Union["str", 1]'},
             ['str'] if sys.version_info >= (3, 11) else [], ''),
            ({'return': 'typing.Optional[str]'}, ['NoneType', 'str'], ''),
            ({'return': 'typing.Optional[str, int]'}, [], ''),  # Takes only one arg
            ({'return': 'typing.Any'},
             ['_AnyMeta'] if sys.version_info >= (3, 11) else [], ''),
    
            ({'return': 'typing.Tuple[int, str]'},
             ['Tuple' if sys.version_info[:2] == (3, 6) else 'tuple'], ''),
            ({'return': 'typing.Tuple[int, str]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.Tuple[int, str]'}, ['str'], 'x()[1]'),
            ({'return': 'typing.Tuple[int, str]'}, [], 'x()[2]'),
    
            ({'return': 'typing.List'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.List[int, str]'}, [], 'x()[0]'),
    
            ({'return': 'typing.Iterator[int]'}, [], 'x()[0]'),
            ({'return': 'typing.Iterator[int]'}, ['int'], 'next(x())'),
            ({'return': 'typing.Iterable[float]'}, ['float'], 'list(x())[0]'),
    
            ({'return': 'decimal.Decimal'}, [], ''),
            ({'return': 'lalalalallalaa'}, [], ''),
            ({'return': 'lalalalallalaa.lala'}, [], ''),
        ]
    )
    def test_string_annotation(annotations, result, code):
        x = lambda foo: 1  # noqa: E731
        x.__annotations__ = annotations
        defs = jedi.Interpreter(code or 'x()', [locals()]).infer()
>       assert [d.name for d in defs] == result
E       AssertionError: assert [] == ['int', 'str']
E         
E         Right contains 2 more items, first extra item: 'int'
E         Use -v to get more diff

test/test_api/test_interpreter.py:697: AssertionError
________________ test_string_annotation[annotations9-result9-] _________________

annotations = {'return': 'typing.Union["str", int]'}, result = ['int', 'str']
code = ''

    @pytest.mark.parametrize(
        'annotations, result, code', [
            ({}, [], ''),
            (None, [], ''),
            ({'asdf': 'str'}, [], ''),
    
            ({'return': 'str'}, ['str'], ''),
            ({'return': 'None'}, ['NoneType'], ''),
            ({'return': 'str().upper'}, [], ''),
            ({'return': 'foo()'}, [], ''),
            ({'return': 'bar()'}, ['float'], ''),
    
            # typing is available via globals.
            ({'return': 'typing.Union[str, int]'}, ['int', 'str'], ''),
            ({'return': 'typing.Union["str", int]'},
             ['int', 'str'] if sys.version_info >= (3, 9) else ['int'], ''),
            ({'return': 'typing.Union["str", 1]'},
             ['str'] if sys.version_info >= (3, 11) else [], ''),
            ({'return': 'typing.Optional[str]'}, ['NoneType', 'str'], ''),
            ({'return': 'typing.Optional[str, int]'}, [], ''),  # Takes only one arg
            ({'return': 'typing.Any'},
             ['_AnyMeta'] if sys.version_info >= (3, 11) else [], ''),
    
            ({'return': 'typing.Tuple[int, str]'},
             ['Tuple' if sys.version_info[:2] == (3, 6) else 'tuple'], ''),
            ({'return': 'typing.Tuple[int, str]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.Tuple[int, str]'}, ['str'], 'x()[1]'),
            ({'return': 'typing.Tuple[int, str]'}, [], 'x()[2]'),
    
            ({'return': 'typing.List'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.List[int, str]'}, [], 'x()[0]'),
    
            ({'return': 'typing.Iterator[int]'}, [], 'x()[0]'),
            ({'return': 'typing.Iterator[int]'}, ['int'], 'next(x())'),
            ({'return': 'typing.Iterable[float]'}, ['float'], 'list(x())[0]'),
    
            ({'return': 'decimal.Decimal'}, [], ''),
            ({'return': 'lalalalallalaa'}, [], ''),
            ({'return': 'lalalalallalaa.lala'}, [], ''),
        ]
    )
    def test_string_annotation(annotations, result, code):
        x = lambda foo: 1  # noqa: E731
        x.__annotations__ = annotations
        defs = jedi.Interpreter(code or 'x()', [locals()]).infer()
>       assert [d.name for d in defs] == result
E       AssertionError: assert [] == ['int', 'str']
E         
E         Right contains 2 more items, first extra item: 'int'
E         Use -v to get more diff

test/test_api/test_interpreter.py:697: AssertionError
_______________ test_string_annotation[annotations10-result10-] ________________

annotations = {'return': 'typing.Union["str", 1]'}, result = ['str'], code = ''

    @pytest.mark.parametrize(
        'annotations, result, code', [
            ({}, [], ''),
            (None, [], ''),
            ({'asdf': 'str'}, [], ''),
    
            ({'return': 'str'}, ['str'], ''),
            ({'return': 'None'}, ['NoneType'], ''),
            ({'return': 'str().upper'}, [], ''),
            ({'return': 'foo()'}, [], ''),
            ({'return': 'bar()'}, ['float'], ''),
    
            # typing is available via globals.
            ({'return': 'typing.Union[str, int]'}, ['int', 'str'], ''),
            ({'return': 'typing.Union["str", int]'},
             ['int', 'str'] if sys.version_info >= (3, 9) else ['int'], ''),
            ({'return': 'typing.Union["str", 1]'},
             ['str'] if sys.version_info >= (3, 11) else [], ''),
            ({'return': 'typing.Optional[str]'}, ['NoneType', 'str'], ''),
            ({'return': 'typing.Optional[str, int]'}, [], ''),  # Takes only one arg
            ({'return': 'typing.Any'},
             ['_AnyMeta'] if sys.version_info >= (3, 11) else [], ''),
    
            ({'return': 'typing.Tuple[int, str]'},
             ['Tuple' if sys.version_info[:2] == (3, 6) else 'tuple'], ''),
            ({'return': 'typing.Tuple[int, str]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.Tuple[int, str]'}, ['str'], 'x()[1]'),
            ({'return': 'typing.Tuple[int, str]'}, [], 'x()[2]'),
    
            ({'return': 'typing.List'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.List[int, str]'}, [], 'x()[0]'),
    
            ({'return': 'typing.Iterator[int]'}, [], 'x()[0]'),
            ({'return': 'typing.Iterator[int]'}, ['int'], 'next(x())'),
            ({'return': 'typing.Iterable[float]'}, ['float'], 'list(x())[0]'),
    
            ({'return': 'decimal.Decimal'}, [], ''),
            ({'return': 'lalalalallalaa'}, [], ''),
            ({'return': 'lalalalallalaa.lala'}, [], ''),
        ]
    )
    def test_string_annotation(annotations, result, code):
        x = lambda foo: 1  # noqa: E731
        x.__annotations__ = annotations
        defs = jedi.Interpreter(code or 'x()', [locals()]).infer()
>       assert [d.name for d in defs] == result
E       AssertionError: assert [] == ['str']
E         
E         Right contains one more item: 'str'
E         Use -v to get more diff

test/test_api/test_interpreter.py:697: AssertionError
_______________ test_string_annotation[annotations11-result11-] ________________

annotations = {'return': 'typing.Optional[str]'}, result = ['NoneType', 'str']
code = ''

    @pytest.mark.parametrize(
        'annotations, result, code', [
            ({}, [], ''),
            (None, [], ''),
            ({'asdf': 'str'}, [], ''),
    
            ({'return': 'str'}, ['str'], ''),
            ({'return': 'None'}, ['NoneType'], ''),
            ({'return': 'str().upper'}, [], ''),
            ({'return': 'foo()'}, [], ''),
            ({'return': 'bar()'}, ['float'], ''),
    
            # typing is available via globals.
            ({'return': 'typing.Union[str, int]'}, ['int', 'str'], ''),
            ({'return': 'typing.Union["str", int]'},
             ['int', 'str'] if sys.version_info >= (3, 9) else ['int'], ''),
            ({'return': 'typing.Union["str", 1]'},
             ['str'] if sys.version_info >= (3, 11) else [], ''),
            ({'return': 'typing.Optional[str]'}, ['NoneType', 'str'], ''),
            ({'return': 'typing.Optional[str, int]'}, [], ''),  # Takes only one arg
            ({'return': 'typing.Any'},
             ['_AnyMeta'] if sys.version_info >= (3, 11) else [], ''),
    
            ({'return': 'typing.Tuple[int, str]'},
             ['Tuple' if sys.version_info[:2] == (3, 6) else 'tuple'], ''),
            ({'return': 'typing.Tuple[int, str]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.Tuple[int, str]'}, ['str'], 'x()[1]'),
            ({'return': 'typing.Tuple[int, str]'}, [], 'x()[2]'),
    
            ({'return': 'typing.List'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['list'], 'list'),
            ({'return': 'typing.List[int]'}, ['int'], 'x()[0]'),
            ({'return': 'typing.List[int, str]'}, [], 'x()[0]'),
    
            ({'return': 'typing.Iterator[int]'}, [], 'x()[0]'),
            ({'return': 'typing.Iterator[int]'}, ['int'], 'next(x())'),
            ({'return': 'typing.Iterable[float]'}, ['float'], 'list(x())[0]'),
    
            ({'return': 'decimal.Decimal'}, [], ''),
            ({'return': 'lalalalallalaa'}, [], ''),
            ({'return': 'lalalalallalaa.lala'}, [], ''),
        ]
    )
    def test_string_annotation(annotations, result, code):
        x = lambda foo: 1  # noqa: E731
        x.__annotations__ = annotations
        defs = jedi.Interpreter(code or 'x()', [locals()]).infer()
>       assert [d.name for d in defs] == result
E       AssertionError: assert [] == ['NoneType', 'str']
E         
E         Right contains 2 more items, first extra item: 'NoneType'
E         Use -v to get more diff

test/test_api/test_interpreter.py:697: AssertionError
=========================== short test summary info ============================
FAILED test/test_api/test_interpreter.py::test_string_annotation[annotations8-result8-]
FAILED test/test_api/test_interpreter.py::test_string_annotation[annotations9-result9-]
FAILED test/test_api/test_interpreter.py::test_string_annotation[annotations10-result10-]
FAILED test/test_api/test_interpreter.py::test_string_annotation[annotations11-result11-]
==== 4 failed, 3770 passed, 22 skipped, 96 deselected, 1 xfailed in 46.74s =====
```

### PR DESCRIPTION
**Pull request by frenzymadness**
In Python 3.14, typing.Union changed its repr from 'typing.Union[X, Y]' to 'X | Y' (PEP 604), breaking annotation inference.

Changes:
- Use getattr() instead of safe_getattr() for __module__ retrieval (getattr_static fails on Union types in Python 3.14)
- Add fallback to typing.get_origin() when regex fails to match
- Normalize Union display back to 'Union[X, Y]' format for consistency
- Update test expectations for invalid annotation edge case in 3.14

Fixes: https://github.com/davidhalter/jedi/issues/2064

It works for me on Python 3.12, 3.13, and 3.14.

### CODE REVIEW

**Comment by frenzymadness** [comment]
The problem in the docs job is caused by ` ModuleNotFoundError: No module named 'pkg_resources' ` which is very likely caused by a newest setuptools being installed.

**Comment by frenzymadness** [comment]
Would you accept fixes for 3.15 here as well?

File: jedi/inference/compiled/access.py [hunk 1]
```diff
@@ -477,22 +477,49 @@ def get_annotation_name_and_args(self):
         """
         name = None
         args = ()
-        if safe_getattr(self._obj, '__module__', default='') == 'typing':
+        # Use getattr instead of safe_getattr for __module__ as getattr_static
+        # fails on typing types in Python 3.14+
+        module = getattr(self._obj, '__module__', '')
```
Comment (davidhalter) [inline jedi/inference/compiled/access.py:482]: Hmm I need to look at this one in detail. I want to understand why safe_getattr doesn't work. We might need some fixes. Also safe_getattr solves a few other problems of code executions that I would want to avoid, but since it's only accessing `__module__` it might be mostly fine (except for `__getattr__` cases).

DO you know why `__module__` is not working anymore?

**Review by davidhalter — COMMENTED** [review]
Thanks! Sorry for the review delay. I think the changes are mostly fine, but I want to review something manually.

**Comment by davidhalter** [comment]
I don't mind fixing things for 3.15 in here.

File: jedi/inference/compiled/access.py (hunk 1 above)
Comment (frenzymadness) [reply jedi/inference/compiled/access.py:482]: I already kinda lost the context here, but AFAIR: in Python 3.14, the typing module's internal implementation changed. The `__module__` attribute on Union types is no longer accessible through the static inspection mechanism that getattr_static() uses.

**Comment by frenzymadness** [comment]
I've pushed the commit I had prepared for 3.15.

File: test/test_api/test_interpreter.py [hunk 2]
```diff
@@ -663,8 +663,11 @@ def bar():
         ({'return': 'typing.Union[str, int]'}, ['int', 'str'], ''),
         ({'return': 'typing.Union["str", int]'},
          ['int', 'str'] if sys.version_info >= (3, 9) else ['int'], ''),
-        ({'return': 'typing.Union["str", 1]'},
-         ['str'] if sys.version_info >= (3, 11) else [], ''),
+        (
+            {'return': 'typing.Union["str", 1]'},
+            ['str'] if (3, 11) <= sys.version_info < (3, 14) else [],
+            '',
+        ),
```
Comment (davidhalter) [inline test/test_api/test_interpreter.py]:
```
>>> import typing
>>> x = typing.Union["str", 1]
>>> x.__args__
(ForwardRef('str'), 1)
```

Also a small nit: I generally prefer to keep the formatting as is. Especially here where all the other cases have this style. I don't think I would have said anything if I didn't look at this in more detail (Might also be an auto formatter, that did this, but still)

File: jedi/inference/compiled/access.py (hunk 1 above)
Comment (davidhalter) [reply jedi/inference/compiled/access.py:482]: Hmm that's fine. I guess I'll take a stab at it then. This function seems cursed anyway. Why are we matching with regex? Nobody seems to know anymore. This probably comes from a time where some of the typing API wasn't available or very different.

**Review by davidhalter — COMMENTED** [review]
I guess probably just revert the test change that makes it worse for 3.14 and then I can try to fix that one. I don't think it will take too much time.

File: test/test_api/test_interpreter.py (hunk 2 above)
Comment (frenzymadness) [reply test/test_api/test_interpreter.py]: There is a job in the CI here running flake8 and that failed with:

```
test/test_api/test_interpreter.py:667:101: E501 line too long (102 > 100 characters)
```

I can revert the commit if you are okay with merging PRs with failed tests like this.

Comment (davidhalter) [reply test/test_api/test_interpreter.py]: No worries, I'll just build upon this PR

Comment (davidhalter) [reply test/test_api/test_interpreter.py]: @frenzymadness Are you going to revert that specific commit so I can work on this?

Comment (frenzymadness) [reply test/test_api/test_interpreter.py]: Droped, force-pushed

**Comment by davidhalter** [comment]
I fixed a few of the remaining issues. Thanks for doing the initial work! Release will hopefully follow soon.
